sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -6267,11 +6274,16 @@ static pi_result USMFreeHelper(pi_context Context, void *Ptr) {
   }
 
   if (ZeDeviceHandle) {
-    // All devices in the context are of the same platform.
-    auto Platform = Context->Devices[0]->Platform;
-    auto Device = Platform->getDeviceFromNativeHandle(ZeDeviceHandle);
-
-    PI_ASSERT(Device, PI_INVALID_DEVICE);
+    pi_device Device;
+    if (Context->Devices.size() == 1) {
+      Device = Context->Devices[0];
+      PI_ASSERT(Device->ZeDevice == ZeDeviceHandle, PI_INVALID_DEVICE);
+    } else {
+      // All devices in the context are of the same platform.
+      auto Platform = Context->Devices[0]->Platform;
+      Device = Platform->getDeviceFromNativeHandle(ZeDeviceHandle);
+      PI_ASSERT(Device, PI_INVALID_DEVICE);
+    }
```

Comment:
Could you explain why this change solved the memory free problem in the sub-device context?

Reply:
I cases where context has exactly one PI device(most of the times), we don't need to search for it in the cache. So this code avoids calling "broken" search, if it can be avoided.

Reply:
Thanks.